tests/e2e/default/contacts/contact-details.wdio-spec.js

```diff
@@ -2,131 +2,120 @@ const commonElements = require('@page-objects/default/common/common.wdio.page.js
-    const role = 'notchw';
-    const parent = placeFactory.place().build({ _id: 'dist1', type: 'district_hospital' });
-    const user = userFactory.build({ username: 'offlineuser', roles: [role] });
-    const patient = personFactory.build({ parent: { _id: user.place._id, parent: { _id: parent._id } } });
-
-    const reports = Array
-      .from({ length: 60 })
-      .map(() => reportFactory.report().build(
-        { form: 'pregnancy_danger_sign' },
-        {
-          patient,
-          submitter: user.contact,
-          fields: { t_danger_signs_referral_follow_up: 'yes' },
-        }
-      ));
-
-    const pregnancyReport = pregnancyFactory.build({
-      fields: {
-        patient_id: patient._id,
-        patient_uuid: patient._id,
-        patient_name: patient.name,
-      },
-    });
-
-    const updatePermissions = async (role, addPermissions, removePermissions = []) => {
-      const settings = await utils.getSettings();
-      settings.roles[role] = { offline: true };
-      addPermissions.map(permission => settings.permissions[permission].push(role));
-      removePermissions.forEach(permission => {
-        settings.permissions[permission] = settings.permissions[permission].filter(r => r !== role);
-      });
-      await updateSettings({ roles: settings.roles, permissions: settings.permissions });
-    };
-
-    const DOCS_DISPLAY_LIMIT = 50;
-
-    before(async () => {
-      const permissions = ['can_view_contacts', 'can_view_contacts_tab', 'can_view_reports', 'can_view_tasks'];
-      await updatePermissions(role, permissions);
-
-      await utils.saveDocs([parent, patient]);
-      await utils.saveDocs([...reports, pregnancyReport]);
-
-      await utils.createUsers([user]);
+describe('Contact details page. Permissions to show reports and tasks', () => {
+  const docsDisplayLimit = 50;
+  const role = 'notchw';
+  const parent = placeFactory.place().build({ _id: 'dist1', type: 'district_hospital' });
+  const user = userFactory.build({ username: 'offlineuser', roles: [role] });
+  const patient = personFactory.build({ parent: { _id: user.place._id, parent: { _id: parent._id } } });
+
+  const updateSettings = async (settings) => {
+    await utils.revertSettings(true);
+    await utils.updateSettings(settings, true);
+  };
+
+  const waitForContactLoaded = async (expectTasks) => {
+    await commonElements.waitForPageLoaded();
+    await contactPage.waitForContactLoaded();
+    if (expectTasks) {
+      // contact loaded only waits for contact summary (which requires reports)
+      // tasks are loaded afterwards and there is no visual indication of them being loaded, or still loading
+      return browser.waitUntil(async () => await (await contactPage.rhsTaskListElement()).isDisplayed());
+    }
+
+    // if we expect _not_ to see tasks, wait so we make sure they have enough time to not appear
+    return browser.pause(1000);
+  };
+
+  const reports = Array
```

Comment:
Do you think that is a good idea to leave the 2 `describe`? I was trying to simplify the test by just having 1 `describe`. Or are you referring to a different part of the test? 😅

Reply:
It's just an example for the case of having nested `describe`, not suggesting to have it in this file. 

Having nested `describe` is useful for 
- Having many test cases about 1 feature allows the dev to collapse the code block for the whole feature in one click. You'll see this more in unit tests, because they usually have more test cases for each unit of code vs e2e might not have that many. 
- Printing the report cleaner and more organized without adding prefixes -- especially because prefixes make the test case description longer and might not fit in one line.  

```
Contact List
     Delete feature
         should not delete if user is admin
         should not delete if it's user's home place
         should not delete if facility has descendants
         should delete when .....   
         should delete when .....   
         should delete when .....   

VS with prefix:

Contact List
     Delete feature - should not delete if user is admin
     Delete feature - should not delete if it's user's home place
     Delete feature - should not delete if facility has descendants
     Delete feature - should delete when .....   
     Delete feature - should delete when .....   
     Delete feature - should delete when .....   
```

- Helps scoping test data or test setup to reduce complexity in the code. For example, by adding some code in the before/beforeEach and after/afterEach without affecting other test cases. 

```
  describe(contact list 
      before(
         login as chw
         Do some setup 
         Prepper some data 
      )
     
      it(should search contact by......)
      
      it(should remove search when....)

       describe(Bulk delete feature 
             before(
                 login as suppervisor 
                 insert 20 contacts for bulk delete.
             )

            it(should delete 20 contact when......)
      )
      
        // This test doesn't get any from the `Bulk delete feature `'s `before` function 
        it(should edit contact when....)

  )

```